### PR TITLE
IS-1305 Initial version of configuration document

### PR DESCRIPTION
**Pull request by Artemkaaas**
Signed-off-by: artem.ivanov <artem.ivanov@dsr-company.com>

### CODE REVIEW

**Comment by jovfer** [comment]
I would like to have review on this ticket from someone outside core team
@esplinr @kdenhartog @mikelodder7 @AxelNennker @swcurran @ianco @brentzundel Please take a look on this PR if you have a chance.

**Comment by swcurran** [comment]
Looks good - Axel has put in some good tweaks.  I would suggest repeating in configuration.md what was said in the cli/README.md  about where the configuration goes. On first pass I just read the configuration.md document and wasn't sure.  Not sure if there is a default config in the repo,  but a link to that would be good.

File: docs/configuration.md [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+```
+{
+  "id": string, Identifier of the wallet.
+        Configured storage uses this identifier to lookup exact wallet data placement.
+  "storage_type": string (optional), Type of the wallet storage. Defaults to 'default'.
+                 'Default' storage type allows to store wallet data in the local file.
+                 Custom storage types can be registered with indy_register_wallet_storage call.
+  "storage_config": object (optional), Storage configuration json. Storage type defines set of supported keys.
+                    Can be optional if storage supports default configuration.
+                    For 'default' storage type configuration is:
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
                    Configuration for 'default' storage type:
```

File: docs/configuration.md [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+```
+
+* Wallet Credentials
+```
+{
+  "key": string, Key or passphrase used for wallet key derivation.
+                 Look to key_derivation_method param for information about supported key derivation methods.
+  "storage_credentials": optional<object> Credentials for wallet storage. Storage type defines set of supported keys.
+                         Can be optional if storage supports default configuration.
+                         For 'default' storage type should be empty.
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
                         Should be empty for 'default' storage type.
```

File: docs/configuration.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  "key_derivation_method": optional<string> Algorithm to use for wallet key derivation:
+                         ARGON2I_MOD - derive secured wallet master key (used by default)
+                         ARGON2I_INT - derive secured wallet master key (less secured but faster)
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
                         ARGON2I_INT - derive secured wallet master key (less secure but faster)
```

File: docs/configuration.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                         RAW - raw wallet key master provided (skip derivation).
+                               RAW keys can be generated with indy_generate_wallet_key call
+}
+```
+
+#### Payment
+
+Libindy provides generic API for building of payment-related transactions.
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
Libindy provides a generic API for building payment-related transactions. 
```

File: docs/configuration.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                         ARGON2I_INT - derive secured wallet master key (less secured but faster)
+                         RAW - raw wallet key master provided (skip derivation).
+                               RAW keys can be generated with indy_generate_wallet_key call
+}
+```
+
+#### Payment
+
+Libindy provides generic API for building of payment-related transactions. 
+These functions look at registered payment methods and call corresponded handlers.
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
These functions look at registered payment methods and call corresponding handlers.
```

File: docs/configuration.md [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+Out-of-box Libindy doesn't not provide support of any payment method, but there is `indy_register_payment_method` API function to register a payment method.
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
Libindy doesn't provide default support of any payment method. There is `indy_register_payment_method` API function to register a payment method.
```

File: docs/configuration.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+So any payment method like Sovrin tokens, Bitcoin tokens, Visa and etc can be plugged into Libindy.
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
Any payment method (e.g., Sovrin tokens, Bitcoin, Visa, etc.) may be added to Libindy through plugins.
```

File: docs/configuration.md [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+
+#### Logging
+Libindy provides two ways on Logger initialization:
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
Libindy provides two options for Logger initialization:
```

File: docs/configuration.md [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+{
+    "crypto_thread_pool_size": int (optional) - size of thread pool for the most expensive crypto operations like keys generation. (4 by default)
+    "collect_backtrace": Optional<bool> - whether errors backtrace should be collected.
+        Capturing of backtrace can affect library performance.
+        NOTE: must be set before invocation of any other API functions.
+}
+```
+
+## Indy-CLI
+There is Command Line Interface (CLI) built over Libindy which provides a set of commands to:
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
There is a Command Line Interface (CLI) built over Libindy which provides a set of commands to:
```

File: docs/configuration.md [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+* Manage wallets
+* Manage pool configurations
+* Manage DIDs
+* Sending transactions to ledger
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
* Send transactions to the ledger
```

File: docs/configuration.md [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+#### Options
+* -h and --help - Print usage. (usage: `indy-cli --help`)
+
+* --logger-config - Init logger according to a config file. <br>
+                    Indy Cli uses [log4rs](https://crates.io/crates/log4rs) logging framework <br>
+                    Usage: `indy-cli --logger-config <path-to-config-file>` <br>
+                    Example: `indy-cli --logger-config logger.yml` <br>
+                    An example of config file can be found [here](../cli/logger.yml) <br>
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
                    An example config file can be found [here](../cli/logger.yml) <br>
```

File: docs/configuration.md [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+                    By default no logger initialized.
+            
+* --plugins - Load custom plugins in Libindy like wallet storage type, payment or else. <br>
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
* --plugins - Load custom plugins in Libindy (e.g., wallet storage type, payment, etc.) <br>
```

File: docs/configuration.md [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+              Usage: `indy-cli --plugins <lib-1-name>:<init-func-1-name>,...,<lib-n-name>:<init-func-n-name>)` <br>
+              Example: `indy-cli --plugins libnullpay.so:nullpay_init`
+
+* --config - Define config file for CLI initialization. <br>
+             Usage: `indy-cli --config <path-to-config-json-file>` <br>
+             Example: `indy-cli --config linux-sample-config.json`
+
+#### Config
+Indy-CLI supports initialization with predefined config file.
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
Indy-CLI supports initialization with a predefined config file. 
```

File: docs/configuration.md [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+Indy-CLI supports initialization with predefined config file. 
+A config file can contain the following fields:
+```
+{
+  "loggerConfig": string, - path to a logger config file (is equal to usage of "--logger-config" option)
+  "plugins": string, - a list of plugins to load in Libindy (is equal to usage of "--plugins" option)
+  "taaAcceptanceMechanism": string - transaction author agreement acceptance mechanism to use for sending write transactions to the Ledger.
+}
+```
+An example of config file can be found [here](../cli/linux-sample-config.json)
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
An example config file can be found [here](../cli/linux-sample-config.json)
```

File: docs/configuration.md [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+
+#### Execution mode
+Indy-CLI will support 2 execution modes:
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
Indy-CLI supports two execution modes:
```

File: docs/configuration.md [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+* Interactive. In this mode CLI will read commands from terminal interactively (command by command).
+* Batch. In this code all commands will be read from file or pipe and executed in series. <br>
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
* Batch. In this mode all commands will be read from file or pipe and executed in series. <br>
```

File: docs/configuration.md [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+Usage: `indy-cli <path-to-file>` <br>
+An example of file congaing batch script:
```
Comment (brentzundel) [inline docs/configuration.md]:
```suggestion
An example of a batch script:
```

**Comment by Artemkaaas** [comment]
Thank @AxelNennker and @brentzundel for corrections!

**Pull request merged**